Always pass the current user into representers

NobodysNightmare opened the pull request:
## OpenProject work packages
- was written for https://community.openproject.org/work_packages/21386
- accidentially also fixes https://community.openproject.org/work_packages/21389
## Description

It was discovered that I forgot to pass the `current_user` into the `AttachmentRepresenter`. Upon further exploration, this error happened quite often and most representers either expect a user anyway or at least could legitimately do that in the future.

I therefore decided to enforce that the `current_user` is passed, for the following benefits:
- less error prone (we can't forget to pass the user, no more soft failures on permission check)
- common interface for all representers
- we can now always pass a `current_user`, allowing us to embed `versions` using the `linked_property` helper
## Related Plugin PRs

:warning: Merge together with: https://github.com/finnlabs/openproject-costs/pull/162

NobodysNightmare committed:
Pass current_user into all AttachmentRepresenters

NobodysNightmare committed:
Enforce passing of current_user into representers

Thereby permission checks will no longer silently fail if no user was
passed.
This uncovered a lot of places where we actually expected a user to be passed, when it wasn't.
This was especially true for collections.
The basic idea is, that virtually all representers could change their behaviour based on the current user.
Representers where this is not the case (e.g. Digest), can still explicitly opt-out.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-inject-current-user.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-inject-current-user.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
I will have to see whether travis accepts my changes on first try. I ran the unit tests locally, but it is a first for the integrative request specs.

NobodysNightmare committed:
Promote ::create to a common interface method

This one is already used by the representers that require a custom field injector.
However, by having a separate method this also means collections (or the new DecoratorFactory)
need to take care and call the ::create method instead of ::new.

I therefore made create the default interface to create representers, which
is now used by Factories and Collections.
Note: I did not yet change all representer instantiations to ::create.
I hope this will happen gradually.

NobodysNightmare commented:
Truns out his will also fix https://community.openproject.org/work_packages/21389 which broke because of a similar mistake.

NobodysNightmare committed:
Fix (untested) relation representer

This one still is not very APIv3, but there ain't no time to improve it...

The angular frontend also seems to rely on the quirks with the relation type changing depending which work package embeds the relation.

Azure7111 commented:
Just a reminder: You might want to check if any plug-in breaks with these changes.

Azure7111 commented on lib/api/v3/string_objects/string_object_representer.rb:
```diff
@@ -34,6 +34,11 @@ module API
       class StringObjectRepresenter < ::API::Decorators::Single
+        # accepting current_user as argument to match common interface, we do not need it though
+        def initialize(string, current_user: nil)
+          super(string, current_user: nil)
```
You might want to apply @houndci's remark here.

NobodysNightmare replied:
As discussed the comment above the method already explains my intention.

Since this is a named parameter I also have no opportunity to rename it to `_current_user`

Azure7111 commented on lib/api/v3/users/user_representer.rb:
```diff
@@ -112,13 +122,13 @@ def _type
 
         def current_user_is_admin
-          current_user && current_user.admin?
+          current_user.admin?
```
How can you be so sure that ``current_user`` is not ``nil`` here?

Azure7111 commented on spec/lib/api/v3/projects/project_collection_representer_spec.rb:
```diff
@@ -31,7 +31,10 @@
-  let(:representer) { described_class.new(projects, self_link) }
+  let(:current_user) { FactoryGirl.build(:user) }
+  let(:representer) {
+    described_class.new(projects, self_link, current_user: current_user)
```
Why is this user no `double`?

NobodysNightmare replied:
Because it is being asked for permissions (`:allowed_to?`)

NobodysNightmare replied:
I only used doubles, where no apropriate user was required.

Azure7111 commented:
Due to a 500 I cannot test the second work package (21389). Here is the stacktrace for that:

```
NoMethodError (undefined method `assignable_versions' for #<API::V3::WorkPackages::Schema::TypedWorkPackageSchema:0x0000001004c508>):
  lib/api/v3/utilities/custom_field_injector.rb:178:in `block in inject_version_schema'
  lib/api/decorators/schema.rb:116:in `instance_exec'
  lib/api/decorators/schema.rb:116:in `block in schema_with_allowed_collection'
  lib/api/root.rb:45:in `call'
  app/middleware/params_parser_with_exclusion.rb:38:in `call'


  Rendered /home/user/.rvm/gems/ruby-2.1.4/gems/actionpack-4.1.13/lib/action_dispatch/middleware/templates/rescues/_source.erb (0.6ms)
  Rendered /home/user/.rvm/gems/ruby-2.1.4/gems/actionpack-4.1.13/lib/action_dispatch/middleware/templates/rescues/_trace.text.erb (0.6ms)
  Rendered /home/user/.rvm/gems/ruby-2.1.4/gems/actionpack-4.1.13/lib/action_dispatch/middleware/templates/rescues/_request_and_response.text.erb (0.9ms)
  Rendered /home/user/.rvm/gems/ruby-2.1.4/gems/actionpack-4.1.13/lib/action_dispatch/middleware/templates/rescues/diagnostics.erb (32.3ms)
```

Azure7111 commented:
As it turns out this was only due to inconsistent data in my db.

Azure7111 commented:
LGTM :+1:

Let me just have a look at the related costs PR before I merge this one.

Azure7111 commented:
Will merge if Travis agrees

Azure7111 commented:
Merging to avoid problems with the already merged costs PR linked to this.
